src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,131 @@
+mod event;
+mod notification;
+
+use event::LogEvent;
+
+use crate::api::{
+    tap::{TapController, TapNotification, TapResult, TapSink},
+    ControlSender,
+};
+use async_graphql::{validators::IntRange, Context, Subscription, Union};
+use futures::StreamExt;
+use itertools::Itertools;
+use tokio::{select, stream::Stream, sync::mpsc, time};
+
+#[derive(Union)]
+/// Log event result which can be a log event or notification
+pub enum LogEventResult {
+    /// Log event
+    LogEvent(event::LogEvent),
+    /// Notification
+    Notification(notification::LogEventNotification),
+}
+
+/// Convert an `api::TapResult` to the equivalent GraphQL type.
+impl From<TapResult> for LogEventResult {
+    fn from(t: TapResult) -> Self {
+        use notification::{LogEventNotification, LogEventNotificationType};
+
+        match t {
+            TapResult::LogEvent(name, ev) => Self::LogEvent(LogEvent::new(&name, ev)),
+            TapResult::Notification(name, n) => match n {
+                TapNotification::Matched => Self::Notification(LogEventNotification::new(
+                    &name,
+                    LogEventNotificationType::Matched,
+                )),
+                TapNotification::NotMatched => Self::Notification(LogEventNotification::new(
+                    &name,
+                    LogEventNotificationType::NotMatched,
+                )),
+            },
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct LogEventsSubscription;
+
+#[Subscription]
+impl LogEventsSubscription {
+    /// A stream of log events emitted from matched component(s)
+    pub async fn output_log_events<'a>(
+        &'a self,
+        ctx: &'a Context<'a>,
+        component_names: Vec<String>,
+        #[graphql(default = 500)] interval: i32,
+        #[graphql(default = 100, validator(IntRange(min = "1", max = "10_000")))] limit: i32,
+    ) -> impl Stream<Item = Vec<LogEventResult>> + 'a {
+        let control_tx = ctx.data_unchecked::<ControlSender>().clone();
+        create_log_events_stream(
+            control_tx,
+            &component_names,
+            // GraphQL only supports 32 bit ints out-the-box due to JSON limitations; we're
+            // casting here to separate concerns and avoid 64 bit scalar deserialization.
+            interval as u64,
+            limit as usize,
+        )
+    }
+}
+
+/// Creates a log events stream based on component names, and a provided interval. Will emit
+/// control messages that bubble up the application if the sink goes away. The stream contains
+/// all matching events; filtering should be done at the caller level.
+fn create_log_events_stream(
+    control_tx: ControlSender,
+    component_names: &[String],
+    interval: u64,
+    limit: usize,
+) -> impl Stream<Item = Vec<LogEventResult>> {
+    let (tx, mut rx) = mpsc::channel(limit);
+    let (mut log_tx, log_rx) = mpsc::channel::<Vec<LogEventResult>>(10);
```

Comment:
Any particular reason 10 was chosen?

Reply:
Only that 1 felt too low, and 100 felt too much 😄 

A value of `1` is too low; `log_tx.send()` would block until the resolver processes the stream, which _might_ mean that the payload is transmitted to the client. I haven't dug too far into the handling of subscription streams in async-graphql to determine whether values are being copied before yielding the next stream value.

10 felt like a sensible default with some margin for back pressure, although this was far from a scientific analysis.

Reply:
Gotcha 😄 Maybe a comment could be added? I could see asking myself this question again when I stumble upon this line.